whatis277/harvest
# bean: Improve http error handling

Adds proper error handling to the http server

The idea is to send http errors with status, message, and redirect URL

Then, log the messages and handle the error throwing in the base handler

For now, this doesn't add an error page/template
I don't think I'm adding that for now

I'm not a fan of how large this PR is
But it was easier than making it a series of PRs

Session testing:
1. Go through the general flow testing from the README 

1. Go through the general flow again but this time:

    1. After logging in, modify the session cookie to something invalid in the browser
    
    ```bash
    > echo "eyJpZCI6ImY4MDY5ZDg1LTQ0NzMtNDgwMC05YzFiLWEyODFjOTczOWMyNCIsInRva2VuIjoiNjZiZjM5NDUtMWQ4MC00NWU3LTkxMTYtODEwMDUzZjFhYzcwIiwiZXhwaXJlc19hdCI6IjIwMjQtMDMtMTRUMTU6Mjc6NTkuNDkwNDIwNDM5WiJ9" | base64 --decode
    {"id":"f8069d85-4473-4800-9c1b-a281c9739c24","token":"66bf3945-1d80-45e7-9116-810053f1ac70","expires_at":"2024-03-14T15:27:59.490420439Z"}
    
    > # modify the token so its not correct
    > echo "{\"id\":\"f8069d85-4473-4800-9c1b-a281c9739c24\",\"token\":\"66bf3945-1d80-45e7-9116-810053f1ac75\",\"expires_at\":\"2024-03-14T15:27:59.490420439Z\"}" | base64
    eyJpZCI6ImY4MDY5ZDg1LTQ0NzMtNDgwMC05YzFiLWEyODFjOTczOWMyNCIsInRva2VuIjoiNjZiZjM5NDUtMWQ4MC00NWU3LTkxMTYtODEwMDUzZjFhYzc1IiwiZXhwaXJlc19hdCI6IjIwMjQtMDMtMTRUMTU6Mjc6NTkuNDkwNDIwNDM5WiJ9Cg==
    ```
    
    1. Refresh the page, ensure you're redirected to `/login` again
    
    1. Check the cookies and ensure your cookie has been cleared
    
    1. Login again, ensure you're redirected back to login

1. Do Step 2 again but this time change the following:

    1. Instead of refreshing that page at 2.2, press on Logout
    
    1. Ensure you're redirected to `/login` again
    
    1. Check the cookies and ensure your cookie has been cleared

Testing base handler:

Generic error testing:
1. Apply the following patch to the code

    ```diff
    diff --git a/bean/internal/adapter/controller/auth/login.go b/bean/internal/adapter/controller/auth/login.go
    index 5942dbd..0e2328d 100644
    --- a/bean/internal/adapter/controller/auth/login.go
    +++ b/bean/internal/adapter/controller/auth/login.go
    @@ -64,6 +64,8 @@ func (c *Controller) renderLogin(
            w http.ResponseWriter,
            data *viewmodel.LoginViewData,
     ) error {
    +       return fmt.Errorf("hello - generic error here")
    +
            err := c.LoginView.Render(w, data)
            if err != nil {
                    return &model.HTTPError{
    
    ```

1. Visit http://whatisbean.local/login

1. Ensure you only see "Something went wrong"

1. Check the docker logs

1. Ensure you see the following

    ```bash
    harvest-bean-1   | server: generic error: hello - generic error here
    ```

1. Revert the patch that was applied

Http error testing:
1. Apply the following patch to the code

    ```diff
    diff --git a/bean/internal/adapter/controller/auth/login.go b/bean/internal/adapter/controller/auth/login.go
    index 5942dbd..3d6e6bf 100644
    --- a/bean/internal/adapter/controller/auth/login.go
    +++ b/bean/internal/adapter/controller/auth/login.go
    @@ -63,7 +63,7 @@ func (c *Controller) LoginForm() model.HTTPHandler {
     func (c *Controller) renderLogin(
            w http.ResponseWriter,
            data *viewmodel.LoginViewData,
    -) error {
    +) *model.HTTPError {
            err := c.LoginView.Render(w, data)
            if err != nil {
                    return &model.HTTPError{
    
    ```

1. Visit http://whatisbean.local/login

1. Check the docker logs

1. Ensure you don't the following

    ```bash
    harvest-bean-1   | server: http error: null
    ```

1. Apply the following patch to the code

    ```diff
    diff --git a/bean/internal/adapter/controller/auth/login.go b/bean/internal/adapter/controller/auth/login.go
    index 5942dbd..34ccd6c 100644
    --- a/bean/internal/adapter/controller/auth/login.go
    +++ b/bean/internal/adapter/controller/auth/login.go
    @@ -63,7 +63,12 @@ func (c *Controller) LoginForm() model.HTTPHandler {
     func (c *Controller) renderLogin(
            w http.ResponseWriter,
            data *viewmodel.LoginViewData,
    -) error {
    +) *model.HTTPError {
    +       return &model.HTTPError{
    +               Status:  http.StatusInternalServerError,
    +               Message: "hello - http error here",
    +       }
    +
            err := c.LoginView.Render(w, data)
            if err != nil {
                    return &model.HTTPError{
    
    ```

1. Visit http://whatisbean.local/login

1. Ensure you only see "Something went wrong"

1. Check the docker logs

1. Ensure you see the following

    ```bash
    harvest-bean-1   | server: http error: {"status":500,"redirect_path":"","message":"hello - http error here"}
    ```
